File: web/src/engine/rendering.ts
import { ARENA_HEIGHT, ARENA_WIDTH, BALL_SIZE, PADDLE_HEIGHT, PADDLE_WIDTH } from "@/engine/config";
import type { GameState } from "@/engine/state";

export function drawGame(ctx: CanvasRenderingContext2D, state: GameState): void {
  ctx.fillStyle = "#101820";
  ctx.fillRect(0, 0, ARENA_WIDTH, ARENA_HEIGHT);

  ctx.strokeStyle = "#4c9f70";
  ctx.lineWidth = 1;
  ctx.setLineDash([2, 2]);
  ctx.beginPath();
  ctx.moveTo(ARENA_WIDTH / 2, 0);
  ctx.lineTo(ARENA_WIDTH / 2, ARENA_HEIGHT);
  ctx.stroke();
  ctx.setLineDash([]);

  ctx.fillStyle = "#f2c14e";
  ctx.fillRect(0, state.paddleLeftY, PADDLE_WIDTH, PADDLE_HEIGHT);
  ctx.fillRect(ARENA_WIDTH - PADDLE_WIDTH, state.paddleRightY, PADDLE_WIDTH, PADDLE_HEIGHT);

  ctx.fillStyle = "#e8505b";
  ctx.fillRect(state.ballX, state.ballY, BALL_SIZE, BALL_SIZE);
}

File: engine/pong_engine/rendering.py
from __future__ import annotations

import numpy as np

from pong_engine.config import (
    ARENA_HEIGHT,
    ARENA_WIDTH,
    BALL_SIZE,
    PADDLE_HEIGHT,
    PADDLE_WIDTH,
)
from pong_engine.state import GameState


def bitmap_from_state(state: GameState) -> np.ndarray:
    bitmap = np.zeros((ARENA_HEIGHT, ARENA_WIDTH), dtype=np.uint8)
    bitmap[
        state.paddle_left_y : state.paddle_left_y + PADDLE_HEIGHT,
        0:PADDLE_WIDTH,
    ] = 1
    bitmap[
        state.paddle_right_y : state.paddle_right_y + PADDLE_HEIGHT,
        ARENA_WIDTH - PADDLE_WIDTH : ARENA_WIDTH,
    ] = 1
    ball_x = int(state.ball_x)
    ball_y = int(state.ball_y)
    bitmap[ball_y : ball_y + BALL_SIZE, ball_x : ball_x + BALL_SIZE] = 1
    return bitmap
